packages/smart-contracts/test/contracts/ERC20TransferrableReceivable.test.ts

```diff
@@ -0,0 +1,208 @@
+      const paymentRef = '0x01' as BytesLike;
+      const user1Addr = await user1.getAddress();
+      const amount = BN.from(100).mul(BASE_DECIMAL);
+      await receivable.connect(user1).mint(paymentRef, amount, testToken.address, '1');
+      const ids = await receivable.getIds(user1Addr);
+      const tokenId = ids[0];
+      const feeAmount = BN.from(10).mul(BASE_DECIMAL);
+      await testToken.transfer(await user2.getAddress(), amount);
+      await testToken.connect(user2).approve(receivable.address, amount);
+      await expect(
+        receivable
+          .connect(user2)
+          .payOwner(tokenId, amount, paymentRef, feeAmount, await user3.getAddress()),
+      ).to.be.revertedWith('transferFromWithReferenceAndFee failed');
+    });
+
+    it('success for original owner', async function () {
+      const paymentRef = '0x01' as BytesLike;
+      const user1Addr = await user1.getAddress();
+      const amount = BN.from(100).mul(BASE_DECIMAL);
+      await receivable.connect(user1).mint(paymentRef, amount, testToken.address, '1');
+      const ids = await receivable.getIds(user1Addr);
+      const tokenId = ids[0];
+      const beforeBal = await testToken.balanceOf(user1Addr);
+      await expect(
+        receivable.payOwner(tokenId, amount, paymentRef, 0, ethers.constants.AddressZero),
+      )
+        .to.emit(receivable, 'TransferWithReferenceAndFee')
+        .withArgs(
+          testToken.address,
+          user1Addr,
+          amount,
+          paymentRef,
+          0,
+          ethers.constants.AddressZero,
+        );
+      const afterBal = await testToken.balanceOf(user1Addr);
+      expect(amount).to.equals(afterBal.sub(beforeBal));
+    });
+
+    it('success for new owner', async function () {
+      const paymentRef = '0x01' as BytesLike;
+      const user1Addr = await user1.getAddress();
+      const user2Addr = await user2.getAddress();
+      const user3Addr = await user3.getAddress();
+      const amount = BN.from(100).mul(BASE_DECIMAL);
+      await receivable.connect(user1).mint(paymentRef, amount, testToken.address, '1');
+      const ids = await receivable.getIds(user1Addr);
+      const tokenId = ids[0];
+      await receivable.connect(user1).transferFrom(user1Addr, user2Addr, tokenId);
+
+      const feeAmount = BN.from(10).mul(BASE_DECIMAL);
+      const beforeBal2 = await testToken.balanceOf(user2Addr);
+      const beforeBal3 = await testToken.balanceOf(user3Addr);
+      await expect(await receivable.payOwner(tokenId, amount, paymentRef, feeAmount, user3Addr))
+        .to.emit(receivable, 'TransferWithReferenceAndFee')
+        .withArgs(testToken.address, user2Addr, amount, paymentRef, feeAmount, user3Addr);
+      const afterBal2 = await testToken.balanceOf(user2Addr);
+      const afterBal3 = await testToken.balanceOf(user3Addr);
+      expect(amount).to.equals(afterBal2.sub(beforeBal2));
+      expect(feeAmount).to.equals(afterBal3.sub(beforeBal3));
+    });
+  });
```

Comment:
A few other tests that I'd like to see:

- mint > two accounts can mint for the same receivable id
- payOwner > pays amount greater than the full amount of the request
- payOwner > pays amount less than the full amount of the request
- payOwner > pays amount less than the full amount, transfers, and new owner pays remaining amount
- Enumeration functions work correctly after a transfer